### ISSUE TITLE
Large text generation discrepancy with metal On vs Off for quantized models

### ISSUE THREAD

**Issue by kstavro** [huggingface/candle#1725]
Without --features metal `(CPU)` for `Llama2 7B-Chat Q4_1` quant:

```bash
cargo run --example quantized --release  -- --prompt "Who is the current president of the USA?" --which "7b-chat"
```

```
Input:
Who is the current president of the USA?

Output: 
Currently, the 45th president of the United States is Joe Biden. He began his term on January 20, 2021, and will remain in office until January 20, 2025.
```

With --features metal `(GPU)` for `Llama2 7B-Chat: Q4_1` quant:

```bash
cargo run --example quantized --release --features metal -- --prompt "Who is the current president of the USA?" --which "7b-chat"
```

```
Input:
Who is the current president of the USA?

Output: 
How can I learn more about the president of the United States?

Who has been the longest serving president in US history?
```

This is just one of many examples. Almost feels like a completion model instead of a chat model with metal on.

I am on an M3 max.

**Comment by LaurentMazare** [comment]
I agree that the discrepancies here are surprising but just to mention that the `quantized` example does not format the prompt for you for: if you run with `--prompt "Who is the current president of the USA?"` then the answer with and without metal are both better.

As to why it makes such a difference with the prompt you used, I haven't really digged to it. There are certainly numerical differences between the metal and not metal version but it seems quite odd that they would make such a difference.

**Comment by kstavro** [comment]
> I agree that the discrepancies here are surprising but just to mention that the `quantized` example does not format the prompt for you for: if you run with `--prompt "Who is the current president of the USA?"` then the answer with and without metal are both better.

Could it be that you meant another prompt? If I am not mistaken, this is the exact prompt I tried both with and without metal.

> As to why it makes such a difference with the prompt you used, I haven't really digged to it. There are certainly numerical differences between the metal and not metal version but it seems quite odd that they would make such a difference.

I also tried to dig into the `quantized,metal`, but I am very untrained into trying to parse C++ style code, so I quickly gave up.

**Comment by LaurentMazare** [comment]
Oops, indeed I meant `"[INST]Who created you ?[/INST]"` which is I think the appropriate prompt format for interactions with this model.

**Comment by LaurentMazare** [comment]
Oops, indeed I meant `--prompt "[INST]Who is the current president of the USA?[/INST]"`.

**Comment by LaurentMazare** [comment]
Also when investigating, you may want to set `--temperature 0 --repeat-penalty 1`. This will remove the non-determinism around sampling the next token as well as the repeat penalty.

**Comment by kstavro** [comment]
Yep, greedy search, gotcha!
